### PR TITLE
add 'code' accessor to QDMDatatype

### PR DESCRIPTION
**Pull request by losborne**
This PR is to add the attribute `code` to the QDM Base Datatype and have it return a single code.

There is an example of this not working on the hmdmmathematica@gmail.com account.  See test case DenExcptPass NoProcSupplyIssueSBEEnc.  Before this change, `.code` will be red.  After the change, it will be green. Unit tests are included, so this was not made a separate JIRA test.

Pull requests into CQL QDM Patient API require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1197
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced. NA
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here:
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here:
- [x] Automated regression test(s) pass

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA test links:
- [x] Justification for using JIRA tests:
- [x] JIRA tests have been added to sprint

**Reviewer 1:**

Name: @holmesie 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] N/A - JIRA tests have been run and pass
- [x] N/A - You agree with the justification for use of JIRA tests or have provided input on why you disagree

**Reviewer 2:**

Name: @mayerm94
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: app/assets/javascripts/datatypes/datatype.js.coffee [hunk 1]
```diff
@@ -42,3 +42,14 @@ class CQL_QDM.QDMDatatype
...
     else
       null
+
+  ###
+  @returns {Code}
+  ###
+  code: ->
+    codeArray = @getCode()
+    if codeArray.length > 0 and codeArray[0].code? and codeArray[0].system?
+      new cql.Code(codeArray[0].code, codeArray[0].system)
```
Comment (holmesie) [inline app/assets/javascripts/datatypes/datatype.js.coffee]: So we're always assuming the first code is the one we want datatypes to return? Just wanted to be clear that is the desired behavior.

File: app/assets/javascripts/datatypes/datatype.js.coffee [hunk 2]
```diff
@@ -42,3 +42,14 @@ class CQL_QDM.QDMDatatype
       return new CQL_QDM.Id(@entry._id)
     else
       null
+
+  ###
+  @returns {Code}
+  ###
+  code: ->
+    codeArray = @getCode()
```
Comment (holmesie) [inline app/assets/javascripts/datatypes/datatype.js.coffee]: I'm not entirely sure how many cycles this would realistically save, but rather than using the getCodes method (which loops over however many codes there are), you could just use `@_codes`. So, I'm thinking something like:

```
code = @_codes[0] if @_codes.length
if code? and code.code? and code.system?
  new cql.Code(code.code, code.system)
```

File: app/assets/javascripts/datatypes/datatype.js.coffee (hunk 1 above)
Comment (losborne) [reply app/assets/javascripts/datatypes/datatype.js.coffee]: Very good question.  There should really only ever be one code, but since bonnie allows for more than one, this is the convention that we have decided to use.  It seems to be correct for all the current cases where `.code` is used, but I'm not entirely sure if there could possibly be a case where an expected code would be later in the array and thus behave incorrectly.  

If that were the case, I do not know of any way to choose one code from the array other than arbitrarily choosing one.  The spec is clear that a single code must be returned, so I think this is as best as we can do without a major change to how bonnie stores codes.

File: app/assets/javascripts/datatypes/datatype.js.coffee (hunk 2 above)
Comment (losborne) [reply app/assets/javascripts/datatypes/datatype.js.coffee]: `@_codes` has this form, which makes things a bit less clean, but the performance improvement should still be there:
image

File: app/assets/javascripts/datatypes/datatype.js.coffee [hunk 3]
```diff
@@ -42,3 +42,16 @@ class CQL_QDM.QDMDatatype
...
     else
       null
+
+  ###
+  @returns {Code}
+  ###
+  code: ->
+    for system, codes of @_codes
+      for code in codes
+        # return the first code of the first system in @_codes
```
Comment (holmesie) [inline app/assets/javascripts/datatypes/datatype.js.coffee]: Ahh, I like this approach even more than the one I suggested!
